### PR TITLE
Tournaments can run on Windows

### PR DESCRIPTION
**Pull request by drvinceknight**
Started with #719 but has a bit more:

- Have added the ability to keep everything in memory: currently this is the default and temporary files are not used at all but I want to think about that some more (might be more efficient to keep the temporary files as default unless we're on windows which we could just detect and in that case keep things in memory). I'm not overly exited with the fix in #713 as it risks keeping files around if anything is interrupted...
- There are some tests for the above that I need to write ideally.
- Have just skipped the parallel processing tests if on windows (have also skipped testing 2.7 and 3.4 on windows because life is short and they were throwing weird bugs). I feel happier to say that parallel processing is not supported on Windows than making hacky changes...

I believe this will pass all CI tests... Heading to bed: will work on some more tomorrow :)

### CODE REVIEW

**Comment by meatballs** [comment]
>  I feel happier to say that parallel processing is not supported on Windows than making hacky changes...

Agreed - we could even wrap the parallel processing in a condition that checks the os

**Comment by drvinceknight** [comment]
> we could even wrap the parallel processing in a condition that checks the os

Good call, will put something like that together now :) :+1:

**Comment by drvinceknight** [comment]
I think this is in good shape for a review now:

- Have added an `in_memory` variable to `tournament.play`. Default is `False`, if `True` it basically just uses an internal interactions dictionary.
- Have added a `axelrod.on_windows` that on loading of the library detects if it's running on a Windows machine.
- If running on a windows machine **and no filename is provided** then `in_memory` will automatically be set to `True`. This is because `NamedTemporaryFiles` are not very useful on Windows (see this known bug: https://bugs.python.org/issue14243).
- I have added test skips, basically dropping support for parallel processing on Windows. There were other bugs that @Ducksual found with parallel processing: I have opened #718 and suggest that that's fixed as a separate issue.

Have added some tests and also a short documentation tutorial about parallel processing.

I suggest that once this is OK'd and in:

- We make a note on #718 to say that once it's fixed we fix the docs which in two places would say that parallel processing is not supported on Windows.
- We open an issue to refactor `ResultSet` so that it's as efficient as `ResultSetFromFile`. Currently it reads the dictionary more often than it needs to. This shouldn't be hard: everything is basically already in place in `ResultSetFromFile`, that just needs to be tweaked so that it worked with a dictionary.

**Comment by meatballs** [comment]
What about the wrapping of the parallel processing in a condition so that it only happens if not windows?

**Comment by drvinceknight** [comment]
> What about the wrapping of the parallel processing in a condition so that it only happens if not windows?

Yup: I missed that (thought I had done it), appveyor just pointed it out. Just pushing that now :)

File: axelrod/__init__.py [hunk 1]
```diff
@@ -1,4 +1,7 @@
 from __future__ import absolute_import
+import sys
+
+on_windows = sys.platform.startswith("win")
```
Comment (meatballs) [inline axelrod/__init__.py]: Any reason you've gone with sys.platform rather than os.name?

Comment (drvinceknight) [inline axelrod/__init__.py]: Not an educated reason: would `os.name` be better? I have to check for `nt` or something like that right?

Comment (meatballs) [inline axelrod/__init__.py]: I've always used ```os.name == 'nt' ``` based on answers at http://stackoverflow.com/questions/4553129/when-to-use-os-name-sys-platform-or-platform-system#11674977

Comment (drvinceknight) [inline axelrod/__init__.py]: Cool: I'll change that 👍

Comment (meatballs) [inline axelrod/__init__.py]: but I wondered if you knew something that I don't !!

**Comment by meatballs** [comment]
Big 👍 from me for this one. That was a painful day....

**Comment by Ducksual** [comment]
I don't know what you're talking about. Windows is easy! ;-)

In all seriousness I tried to be fairly careful to clean up the files but it would have been really nice if NamedTemporaryFile actually did something useful in Windows (or if TemporaryDirectory had existed before Python 3.2).

Multiprocessing on windows is a separate nightmare.

File: appveyor.yml [hunk 2]
```diff
@@ -1,10 +1,11 @@
 environment:
   matrix:
-    - PYTHON: "C:\\Python27"
-    - PYTHON: "C:\\Python34"
+    #- PYTHON: "C:\\Python27"
```
Comment (marcharper) [inline appveyor.yml]: Can we delete these if they won't be used?

File: axelrod/tests/integration/test_tournament.py [hunk 3]
```diff
@@ -34,9 +33,9 @@ def test_full_tournament(self):
         tournament = axelrod.Tournament(name='test', players=strategies,
                                         game=self.game, turns=2,
                                         repetitions=2)
-        tmp_file = tempfile.NamedTemporaryFile()
+        filename = "test_outputs/test_tournament.csv"
```
Comment (marcharper) [inline axelrod/tests/integration/test_tournament.py]: Are we going to run into issues if the directory doesn't exist?

**Comment by drvinceknight** [comment]
> In all seriousness I tried to be fairly careful to clean up the files but it would have been really nice if NamedTemporaryFile actually did something useful in Windows (or if TemporaryDirectory had existed before Python 3.2).

You had done a great job, and it's highly appreciated. You figured out all the bugs 👍 

> 
> Multiprocessing on windows is a separate nightmare.

We'll get there eventually (hopefully)!

Either way hopefully this will all be in place so that you could contribute actual Game Theory next :)

File: axelrod/tournament.py [hunk 4]
```diff
@@ -93,11 +102,15 @@ def play(self, build_results=True, filename=None,
             self.progress_bar = tqdm.tqdm(total=len(self.match_generator),
                                           desc="Playing matches")
 
-        self.setup_output_file(filename)
+        if on_windows and (filename is None):
```
Comment (marcharper) [inline axelrod/tournament.py]: So if the tournament takes up a lot of memory on windows, it won't work?

**Comment by Ducksual** [comment]
I actually checked and NamedDirectory wouldn't have helped (at least in cleaning up the files on windows). As far as I can tell the implementation of NamedTemporaryFile on windows with the locking behaviour is the only way to guarantee the file will be deleted. Without locking something else could open the file and then Windows wouldn't allow it to be removed.

File: axelrod/__init__.py (hunk 1 above)
Comment (drvinceknight) [inline axelrod/__init__.py]: Nope: I just got scared away by the `nt` :) Have changed this with 72238a3

File: axelrod/tests/integration/test_tournament.py (hunk 3 above)
Comment (drvinceknight) [inline axelrod/tests/integration/test_tournament.py]: The directory should exist, I've put a `README.md` file in there so cloning the library should put it there right?

File: appveyor.yml (hunk 2 above)
Comment (drvinceknight) [inline appveyor.yml]: Yup.

File: axelrod/tournament.py (hunk 4 above)
Comment (drvinceknight) [inline axelrod/tournament.py]: Yup: it it takes up a lot of memory then it will crash. It will work fine if a filename is provided though.

**Comment by drvinceknight** [comment]
> I doubt many windows users would use the library without a GUI and it makes the code more complex to maintain

I know of 3 that have (with anaconda, Jupyter etc I actually would hope we would not close the door on windows users). I expect I'll have students doing so too. Important to get it working natively on windows even if it does make it harder for us. 

> 
> Has anyone tested the library with bash on windows? That might be a better way to have some support on windows, and who knows, maybe multiprocessing works?

That's not really running on Windows though. I expect that if it runs on bash it might not pick up the `on_windows` flag? Apparently that's a full ubuntu thing? (I have no idea).

File: appveyor.yml (hunk 2 above)
Comment (drvinceknight) [inline appveyor.yml]: Have pushed that.

**Comment by meatballs** [comment]
> I doubt many windows users would use the library without a GUI 

Several of the students who came to the sprint were on windows - and I suspect that's often the case for young undergrads coming to university with their shiny new laptops!

**Comment by Ducksual** [comment]
Bash on windows is Ubuntu related - but still beta and a little more work to install than Anaconda. I'm not sure if all the Linux features would work on it or not but I'm installing it now so I can find out.

**Comment by drvinceknight** [comment]
> Bash on windows is Ubuntu related - but still beta and a little more work to install than Anaconda. I'm not sure if all the Linux features would work on it or not but I'm installing it now so I can find out.

Thanks @Ducksual: that's helpful. 

Either way I think that in the interest of diversity we should work to offer native support. This PR is a first step there, we can always open more PRs to improve it and reduce the complexity :)

**Comment by drvinceknight** [comment]
> I'm installing it now so I can find out.

@Ducksual it would be interesting to know what the following does on bash for windows...:

```
>>> import os
>>> os.name
```

**Comment by Ducksual** [comment]
>>> import os
>>> os.name
'posix'
>>>

**Comment by drvinceknight** [comment]
Neat :)

**Comment by Ducksual** [comment]
It's kind of neat but a little intimidating if you haven't used linux/bash before. After changing your machine to developer mode and going through the install you still don't have any gui support for linux.

**Comment by drvinceknight** [comment]
> It's kind of neat but a little intimidating if you haven't used linux/bash before.

Oh absolutely. Bash for windows isn't really a solution. I was just curious :)

**Comment by drvinceknight** [comment]
> Bash for windows isn't really a solution.

Although it might be for some people as it looks like you would be able to run parallel tournaments with it.

**Comment by marcharper** [comment]
Thanks @ Ducksual @drvinceknight @meatballs ! Great work everyone.

**Pull request merged**